### PR TITLE
Add initial http/https proxy support

### LINKED ISSUE #884: Add http&https proxy support
it that any plan to Add http&https proxy support?

### PR DESCRIPTION
**Pull request by martingoldstone**
# Description

Pass proxy information to aiohttp from either proxy environment variables or per-connector parameters in configuration.yaml, and updated example config for passing per-connector proxy server information. Matrix connector supports only environment variables, as it seems the actual get/post operations are done in another module outside of this.

Fixes #884 

## Status
**READY** 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

- Running in a docker container, linked to another docker container running squid, set configuration.yaml to point at squid for telegram connector. Observe tcpdump in squid showing port 443 traffic heading to telegram, and observe debug output in opsdroid showing bot name for telegram
- Running in a docker container, linked to another docker container running squid, set https_proxy environment variable to point to squid. Observe tcpdump in squid showing port 443 traffic heading to telegram, and observe debug output in opsdroid showing bot name for telegram

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [] I have added tests that prove my fix is effective or that my feature works
  - Current tests seem to all be offline tests. A test for the proxy code would need to be online, and any proxy configuration would be be different from one instance to the next
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by martingoldstone** [comment]
Not sure how I missed the  os import, resolved now.

 I take your point about there being no value in defining the proxy per connector, I think I got caught up with the idea of providing maximum flexibility, even though it made it more complicated than it needed to be. I've updated this to use trust_env where aiohttp is being used, setting a proxies dict from the environment variables when requests it used, and passing over just the HTTPS_PROXY variable for the slack connector.

**Comment by martingoldstone** [comment]
I did repeat the Telegram testing, yes. I've also tested both Slack and Webex Teams, and observed traffic going through my proxy when the environment variable was set, and direct traffic when there was no HTTP_PROXY environment variable. I had to update the Slack client (as in #1219) as I was getting errors resulting in an unhandled exception, but this happened without my modified code as well.

**Comment by jacobtomlinson** [comment]
Yeah thanks for that, I just merged #1219 so that shouldn't be an issue now.

**Comment by jacobtomlinson** [comment]
Resolved merge conflicts.

File: opsdroid/connector/slack/__init__.py [hunk 1]
```diff
@@ -30,10 +31,16 @@ def __init__(self, config, opsdroid=None):
         self.chat_as_user = config.get("chat-as-user", False)
         self.ssl_context = ssl.create_default_context(cafile=certifi.where())
         self.slack = slack.WebClient(
-            token=self.token, run_async=True, ssl=self.ssl_context
+            token=self.token,
```
Comment (FabioRosado) [inline opsdroid/connector/slack/__init__.py]:
```suggestion
            token=self.token,
```

File: opsdroid/connector/slack/__init__.py [hunk 2]
```diff
@@ -30,10 +31,16 @@ def __init__(self, config, opsdroid=None):
         self.chat_as_user = config.get("chat-as-user", False)
         self.ssl_context = ssl.create_default_context(cafile=certifi.where())
         self.slack = slack.WebClient(
-            token=self.token, run_async=True, ssl=self.ssl_context
+            token=self.token, 
+            run_async=True,
```
Comment (FabioRosado) [inline opsdroid/connector/slack/__init__.py]:
```suggestion
            run_async=True,
```

File: opsdroid/connector/slack/__init__.py [hunk 3]
```diff
@@ -30,10 +31,16 @@ def __init__(self, config, opsdroid=None):
         self.chat_as_user = config.get("chat-as-user", False)
         self.ssl_context = ssl.create_default_context(cafile=certifi.where())
         self.slack = slack.WebClient(
-            token=self.token, run_async=True, ssl=self.ssl_context
+            token=self.token, 
+            run_async=True, 
+            ssl=self.ssl_context,
```
Comment (FabioRosado) [inline opsdroid/connector/slack/__init__.py]:
```suggestion
            ssl=self.ssl_context,
```

File: opsdroid/connector/slack/__init__.py [hunk 4]
```diff
@@ -30,10 +31,16 @@ def __init__(self, config, opsdroid=None):
...
         self.slack = slack.WebClient(
-            token=self.token, run_async=True, ssl=self.ssl_context
+            token=self.token, 
+            run_async=True, 
+            ssl=self.ssl_context, 
+            proxy=os.environ.get("HTTPS_PROXY"),
         )
         self.slack_rtm = slack.RTMClient(
-            token=self.token, run_async=True, ssl=self.ssl_context
+            token=self.token,
```
Comment (FabioRosado) [inline opsdroid/connector/slack/__init__.py]:
```suggestion
            token=self.token,
```

File: opsdroid/connector/slack/__init__.py [hunk 5]
```diff
@@ -30,10 +31,16 @@ def __init__(self, config, opsdroid=None):
...
-            token=self.token, run_async=True, ssl=self.ssl_context
+            token=self.token, 
+            run_async=True, 
+            ssl=self.ssl_context, 
+            proxy=os.environ.get("HTTPS_PROXY"),
         )
         self.slack_rtm = slack.RTMClient(
-            token=self.token, run_async=True, ssl=self.ssl_context
+            token=self.token, 
+            run_async=True,
```
Comment (FabioRosado) [inline opsdroid/connector/slack/__init__.py]:
```suggestion
            run_async=True,
```

File: opsdroid/connector/slack/__init__.py [hunk 6]
```diff
@@ -30,10 +31,16 @@ def __init__(self, config, opsdroid=None):
...
+            token=self.token, 
+            run_async=True, 
+            ssl=self.ssl_context, 
+            proxy=os.environ.get("HTTPS_PROXY"),
         )
         self.slack_rtm = slack.RTMClient(
-            token=self.token, run_async=True, ssl=self.ssl_context
+            token=self.token, 
+            run_async=True, 
+            ssl=self.ssl_context,
```
Comment (FabioRosado) [inline opsdroid/connector/slack/__init__.py]:
```suggestion
            ssl=self.ssl_context,
```

**Comment by jacobtomlinson** [comment]
Closing in favour of #1268

**Pull request closed without merging**